include/databaseFunctions_blog_pic.php

```diff
@@ -1,58 +1,450 @@
+				// die();
+				$_SESSION['username'] = $_REQUEST['username'];
+				$_SESSION['userID'] = GetUser($_REQUEST['username'])['userID'];
+				header("Location: /index.php");
+				exit();
+			}else{
+				// var_dump($errors);
+				foreach($errors as $key=>$error){
+					DisplayError($key, $error);
+				}
+				// if(isset($errors['username'])){
+				// 	echo"<div class='required'>Please enter your username.</div>";
+				// }
+				// if(isset($errors['password'])){
+				// 	echo"<div class='required'>Please enter your password.</div>";
+				// }
+				// if(isset($errors['match'])){
+				// 	echo"<div class='required'>Your username and password do not match.</div>";
+				// }
+				// if(isset($errors['DNE'])){
+				// 	echo"<div class='required'>There is no account associated with that username.</div>";
+				// }
+			}
+		}
+		echo"<form method='post'>";
+		ShowTextField(true, 'username');
+		ShowPasswordField('password', 'password');
+		echo"<input type='submit' name = 'login' value='Login'>
+			</form>";
+	}
+
+	function IsLoggedIn(){
+		echo"";
+		if(isset($_SESSION['userID'])){
+			return true;
+		}
+		return false;
+	}
+
+	function PersonalHeading(){
+		echo"
+			<p class='personal'>hi ".$_SESSION['username']."</p>
+		";
+		Logout();
+	}
+
+	function Logout(){
+		if(isset($_REQUEST['logout'])){
+			// echo"<script>
+			// 		alert('Are you sure you want to log out?')
+			// 	</script>
+			// ";
+			session_destroy();
+			header("Location: /index.php");
+			exit();
+		}
+		echo"
+			<form>
+				<input class='personal' type='submit' name='logout' value='Log Out'>
+			</form>
+		";
+	}
+
+	function ValidDelete($postID){
+		// echo"validdelete called";
+		// var_dump($_SESSION);
+		if(isset($_SESSION['userID'])){
+			if($_SESSION['userID'] == GetPostCreator($postID)){
+				return true;
+			}
+			if(GetUser($_SESSION['username'])['userType'] == 'admin'){
+				return true;
+			}
+		}
+		if(GetPostCreator($postID) == NULL){
+			return true;
+		}
+		return false;
+	}
+
+	//TAG DATABASE FUNCTIONS
+	function DeleteTagFromPost($tagname, $postID){		//CHECK DBQUERY SYNTAX
+		$result = dbQuery("
+			DELETE FROM posttags
+			WHERE (SELECT tagname
+				FROM tags
+				INNER JOIN posttags ON
+				tags.tagID = posttags.tagID) = :tagname
+			AND postID = :postID
+		", array('tagname'=>$tagname, 'postID'=>$postID))->fetch();
+	}
+
+	function DeleteTagOverall($name){
+		$result = dbQuery("
+			DELETE FROM tags
+			WHERE tagname = :name
+		", array('name'=>$name))->fetch();
+	}
+
+	function NewTag($name){
+		dbQuery("
+			INSERT INTO tags (tagname)
+			VALUES(:name)
+		", array('name'=>$name))->fetch();
+	}
+
+	function AttachTags($postID, $tagarray){
+		// echo"attachtags called, postID: ";
+		// var_dump($postID);
+		for($i=0;$i<sizeof($tagarray);$i++){
+			// die("forloop");
+			if(($tagarray[$i] == 'null')||($tagarray[$i] == '')){
+				continue;
+			}
+			if(!TagExists($tagarray[$i])){
+				echo"<br>CREATING A NEW TAG<br>";
```

Comment:
probably don't want this here. Normally after submitting a form you want to redirect the page, but you can't do that if you've echoed something.